addons/purchase_stock/models/account_move_line.py

```diff
@@ -216,34 +156,116 @@ def _generate_price_difference_vals(self, layers):
+                            layers_to_consume.append((layer, qty_to_invoice_per_layer[layer][1]))
+                while float_compare(aml_qty, 0, precision_rounding=self.product_id.uom_id.rounding) > 0 and layers_to_consume:
+                    layer, total_layer_qty_to_invoice = layers_to_consume[0]
+                    layers_to_consume = layers_to_consume[1:]
+                    if float_is_zero(total_layer_qty_to_invoice, precision_rounding=self.product_id.uom_id.rounding):
+                        continue
+                    common_qty = min(aml_qty, total_layer_qty_to_invoice)
+                    aml_qty -= common_qty
+                    qty_to_invoice_per_layer[layer][1] -= sign * common_qty
+                    layers_and_invoices_qties[layer, invoice] = [common_qty, common_qty]
+                    layers_and_invoices_qties[layer, impacted_invoice][1] -= common_qty
+
+        return qty_to_invoice_per_layer, layers_and_invoices_qties
+
+    def _prepare_pdiff_vals(self, layer, aml, layer_price_unit, out_qty_to_invoice, qty_to_correct):
+        svl_vals_list = []
+        aml_vals_list = []
+
+        aml_gross_price_unit = aml._get_gross_unit_price()
+        # convert from aml currency to company currency
+        aml_price_unit = aml_gross_price_unit / aml.currency_rate
+        aml_price_unit = aml.product_uom_id._compute_price(aml_price_unit, self.product_id.uom_id)
+
+        unit_valuation_difference = aml_price_unit - layer_price_unit
+
+        # Generate the AML values for the already out quantities
+        # convert from company currency to aml currency
+        unit_valuation_difference_curr = unit_valuation_difference * self.currency_rate
+        unit_valuation_difference_curr = self.product_id.uom_id._compute_price(unit_valuation_difference_curr, self.product_uom_id)
+        out_qty_to_invoice = self.product_id.uom_id._compute_quantity(out_qty_to_invoice, self.product_uom_id)
+        if (
+            not self.currency_id.is_zero(unit_valuation_difference_curr * out_qty_to_invoice) and
+            self.product_id.valuation == 'real_time'
+        ):
+            aml_vals_list += self._prepare_pdiff_aml_vals(out_qty_to_invoice, unit_valuation_difference_curr)
+
+        # Generate the SVL values for the on hand quantities (and impact the parent layer)
+        po_pu_curr = self.purchase_line_id.currency_id._convert(
+            self.purchase_line_id.price_unit,
+            self.currency_id,
+            self.company_id,
+            self.move_id.invoice_date or self.date or fields.Date.context_today(self),
+            round=False
+        )
+        price_difference_curr = po_pu_curr - aml_gross_price_unit
+        if not float_is_zero(unit_valuation_difference * qty_to_correct, precision_rounding=self.company_id.currency_id.rounding):
+            svl_vals = self._prepare_pdiff_svl_vals(layer, qty_to_correct, unit_valuation_difference, price_difference_curr)
```

Comment:
Previously, it was `self._prepare_pdiff_svl_vals(layer, sign * qty_to_correct, [...])`.
Now you apply the `sign` *before* calling `_prepare_pdiff_vals()`, which means the `qty_to_correct` in other lines in this method, although used the same way as before your change *isn't* the same value as it used to be.
Couldn't it cause issues?

Reply:
I don't think so because the only usage of `qty_to_correct` beside this after the new affectation is `float_is_zero([...] qty_to_correct)` and the sign shouldn't matter there.

Reply:
Indeed, you're right :+1: